src/fabric/src/fabric2_db_expiration.erl

```diff
@@ -0,0 +1,273 @@
+        false ->
+            timer:sleep(100),
+            wait_for_couch_jobs_app()
+    end.
+
+
+maybe_add_job() ->
+    case couch_jobs:get_job_data(undefined, ?JOB_TYPE, job_id()) of
+        {error, not_found} ->
+            Now = erlang:system_time(second),
+            ok = couch_jobs:add(undefined, ?JOB_TYPE, job_id(), #{}, Now);
+        {ok, _JobData} ->
+            ok
+    end.
+
+
+cleanup(false) ->
+    timer:sleep(?CHECK_ENABLED_SEC * 1000),
+    exit(normal);
+
+cleanup(true) ->
+    Now = erlang:system_time(second),
+    ScheduleSec = schedule_sec(),
+    Opts = #{max_sched_time => Now + min(ScheduleSec div 3, 15)},
+    case couch_jobs:accept(?JOB_TYPE, Opts) of
+        % maybe handle timeout here, need to check the api
+        {ok, Job, Data} ->
+            try
+                {ok, Job1, Data1} = process_expirations(Job, Data),
+                ok = resubmit_job(Job1, Data1, schedule_sec())
+            catch
+                _Tag:Error ->
+                    Stack = erlang:get_stacktrace(),
+                    couch_log:error("~p : processing error ~p ~p ~p",
+                        [?MODULE, Job, Error, Stack]),
+                    ok = resubmit_job(Job, Data, ?ERROR_RESCHEDULE_SEC),
+                    exit({job_error, Error, Stack})
+            end;
+        {error, not_found} ->
+            timer:sleep(1000),
+            ?MODULE:cleanup(is_enabled())
+    end.
+
+
+resubmit_job(Job, Data, After) ->
+    Now = erlang:system_time(second),
+    SchedTime = Now + After,
+    couch_jobs_fdb:tx(couch_jobs_fdb:get_jtx(), fun(JTx) ->
+         {ok, Job1} = couch_jobs:resubmit(JTx, Job, SchedTime),
+         ok = couch_jobs:finish(JTx, Job1, Data)
+    end),
+    ok.
+
+
+process_expirations(#{} = Job, #{} = Data) ->
+    Callback = fun(Value, Acc) ->
+        NewAcc = case Value of
+            {meta, _} -> Acc;
+            {row, DbInfo} ->
+                TotalLen = length(Acc),
+                process_row(TotalLen, Acc, DbInfo);
+            complete ->
+                TotalLen = length(Acc),
+                if TotalLen == 0 -> Acc; true ->
+                    [{LastDelete, _, _} | _] = Acc,
+                    TotalLen = length(Acc),
+                    delete_dbs(lists:sublist(Acc, TotalLen - LastDelete)),
+                    Acc
+                end
+        end,
+        {ok, NewAcc}
+    end,
+    {ok, _Infos} = fabric2_db:list_deleted_dbs_info(Callback, [], []),
+    {ok, Job, Data}.
+
+
+process_row(0, _Acc, DbInfo) ->
+    DbName = proplists:get_value(db_name, DbInfo),
+    TimeStamp = proplists:get_value(timestamp, DbInfo),
+    [{0, DbName, TimeStamp}];
+
+process_row(TotalLen, Acc, DbInfo) ->
+    [{LastDelete, _, _} | _] = Acc,
+    NumberToDelete = TotalLen - LastDelete,
+    DeleteBatch = expiration_batch(),
+    LastDelete2 = case NumberToDelete == DeleteBatch of
+        true ->
+            delete_dbs(lists:sublist(Acc, DeleteBatch)),
+            report_progress(TotalLen),
```

Comment:
If we don't do batching, here we could just keep the last updated time accumulator and if more than `?JOB_TIMEOUT_SEC div 2` seconds passed we'd update the progress. Them maybe one last update when we finished the job.